### ISSUE TITLE
Problem with DefaultFragmentStateChanger

### ISSUE THREAD

**Issue by uberchilly** [Zhuinden/simple-stack-extensions#10]
Situation: 
I have a screen key like this `
@Parcelize
data class BusinessPickerScreenKey(val businessId: String = "") : DefaultFragmentKey(),
` representing fragment screen

If I have a history of only this screen with an empty id and I end up calling 
`
backstack.setHistory(
      History.of(BusinessPickerScreenKey(businessIdFromNotification)),
      StateChange.REPLACE
  )
`
I end up in a situation where the previous BusinessPickerFragment is removed but the new one is not added. 
Potentially problematic piece of code
```
     for(DefaultFragmentKey newKey : newKeys) {
          final Fragment fragment = fragmentManager.findFragmentByTag(newKey.getFragmentTag());
          if(newKey.equals(stateChange.topNewKey())) {
              if(fragment != null) {
                  if(fragment.isRemoving()) { // fragments are quirky, they die asynchronously. Ignore if they're still there.
                      fragmentTransaction.replace(containerId, topNewFragment, newKey.getFragmentTag());
                  } else if(isNotShowing(fragment)) {
                      startShowing(fragmentTransaction, fragment);
                  }
              } else {
                  // add the newly created fragment if it's not there yet
                  fragmentTransaction.add(containerId, topNewFragment, newKey.getFragmentTag());
              }
          } else {
              if(fragment != null && !isNotShowing(fragment)) {
                  stopShowing(fragmentTransaction, fragment);
              }
          }
      }
```
And more specifically fragment is not null in this situation and it is not removing and not isNotShowing so it misses all the checks and just does nothing and the app shows a blank screen with no fragments because it is removed by the previous for loop that removes screens.

Not sure if this was intended behavior especially when you call `setHistory` but the user of the library is required to check for this situation manually and to call ` backstack.forceClear()`  before calling `setHistory()`

Btw SimpleStateChanger correctly determines that the new top is not the same as the previous top because of the change in parameter `id` 
```
 if(stateChange.isTopNewKeyEqualToPrevious()) {
    completionCallback.stateChangeComplete();
    return;
 }
```
I use this to implement deep linking in the app from notification so I would expect setHistory to drop previous history and set a new one and not to need to check for the state of the current backstack so I can be sure to navigate correctly. 

So in short if you have a history of one screen called X and you call setHistory() to new history with the same fragment you end up with no fragment at all which looks like a bug

**Comment by Zhuinden** [comment]
Hi @uberchilly,

This actually has come up multiple times in the past ~ while I try to provide a "reasonable default" from `DefaultFragmentKey`, the problem is that I can't by default return `toString()` because it needs to be overridden in order to provide something that is stable (the default implementation of `toString()` is not stable across process death, as it adds `object.hashCode()` to the end, which changes between sessions).

In retrospect, I probably should have just made `DefaultFragmentKey` abstract, and make `getFragmentTag()` abstract.

Currently it uses `getClass().getName()` by default, which is stable, but it won't take your arguments into account.

https://github.com/Zhuinden/simple-stack-extensions/blob/80950c65f700cd1d9620e06e7b2afa289ccb46d3/fragments/src/main/java/com/zhuinden/simplestackextensions/fragments/DefaultFragmentKey.java#L43-L46

So your app will actually work if you override `getFragmentTag()` in `BusinessPickerScreenKey` and make it return `toString()`.

**Comment by Zhuinden** [comment]
@uberchilly have you tried overriding `getFragmentTag()` to return `toString()`?

**Comment by uberchilly** [comment]
@Zhuinden thank you for the quick answer. I will test it tomorrow

**Comment by uberchilly** [comment]
It worked with `toString()`, it removed the old fragment properly and created a new one. What is kind of weird is that it didn't call `bindServices()` again for the new one so it basically reused the old binding, in my case old view model for the new fragment, and the old view model has the old `BusinessPickerScreenKey ` with empty `businessId` instead of the new one which means that this intent was lost.   

Edit1: I probably need to update `getScopeTag()`  to return `toString()` as well
Edit2: Yes everything worked now

**Comment by Zhuinden** [comment]
Sometimes I wonder if I should have made those functions abstract, this question does come up every now and then. That's binary breaking change though so it is what it is now. 

Glad it works  :+1:

**Comment by uberchilly** [comment]
Maybe just provide this situation as an example in documentation.
